c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -417,17 +441,32 @@ struct ExpandableSegment {
       prop.location.id = static_cast<int>(device_);
       auto status =
           DriverAPI::get()->cuMemCreate_(&handle, segment_size_, &prop, 0);
-      if (status == CUDA_ERROR_OUT_OF_MEMORY) {
-        for (auto j : c10::irange(begin, i)) {
-          // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
-          auto h = handles_.at(j).value();
-          handles_.at(j) = std::nullopt;
-          C10_CUDA_DRIVER_CHECK(DriverAPI::get()->cuMemRelease_(h.handle));
+      if (status != CUDA_SUCCESS) {
+        if (status == CUDA_ERROR_OUT_OF_MEMORY) {
+          for (auto j : c10::irange(begin, i)) {
+            // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+            auto h = handles_.at(j).value();
+            handles_.at(j) = std::nullopt;
+            C10_CUDA_DRIVER_CHECK(DriverAPI::get()->cuMemRelease_(h.handle));
+          }
+          trimHandles();
+          return rangeFromHandles(begin, begin);
+        } else if (
+            CUDAAllocatorConfig::expandable_segments_handle_type() ==
+            Expandable_Segments_Handle_Type::FABRIC_HANDLE) {
+          // we are testing if we can use fabric handle.
+          // if we can, we will use it.
+          // if we can't, we will use posix file handle.
+          // so we should not return an error here.
+          // in practice, we can get CUDA_ERROR_NOT_SUPPORTED or
+          // CUDA_ERROR_NOT_PERMITTED to be safe, any non out-of-memory error is
+          // considered as the handle type is not supported. if the handle type
+          // is not supported, return a null range to indicate it.
+          return SegmentRange(nullptr, 0);
```

Comment:
this won't clear the error though, so next error check would return false?
Also what happens if allocation succeeds (it requires just the current driver version iiuc), but subsequent handle exhange fails due to missing permissions? 
@malfet is there a way to check if imex permissions are set correctly?

Reply:
> what happens if allocation succeeds (it requires just the current driver version iiuc)

it requires the imex channel to be set up. so allocation succeeds <==> handle exchange succeeds.

Reply:
i checked the error handling doc https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__ERROR.html , it does not say anything about clearing the error. I think driver API error is just returned, without affecting the following call?

Reply:
i think this is demonstrated by the passing tests. they don't have the imex channels set up, so this driver api call will fail. but later calls are still successful.